### PR TITLE
Add pose graph optimization example

### PR DESCRIPTION
**Pull request by fantaosha**
## Types of changes

- [ ] Docs change / refactoring / dependency upgrade
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist

- [ ] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the **CONTRIBUTING** document.
- [ ] I have completed my CLA (see **CONTRIBUTING**)
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

### CODE REVIEW

File: theseus/utils/examples/__init__.py [hunk 1]
```diff
@@ -3,26 +3,47 @@
...
+    from .motion_planning import (
+        InitialTrajectoryModel,
+        MotionPlanner,
+        ScalarCollisionWeightAndCostEpstModel,
+        ScalarCollisionWeightModel,
+        TrajectoryDataset,
+        generate_trajectory_figs,
+    )
+except ModuleNotFoundError:
+    print(
```
Comment (mhmukadam) [inline theseus/utils/examples/__init__.py:16]: This print was replaced to warning in #119. Might make sense to rebase on main after #119 is merged.

File: theseus/utils/examples/pose_graph/pose_prior_error.py [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+# LICENSE file in the root directory of this source tree.
+
+from typing import List, Optional, Tuple
+
+import torch
+
+import theseus as th
+
+
+class PosePirorError(th.CostFunction):
```
Comment (mhmukadam) [inline theseus/utils/examples/pose_graph/pose_prior_error.py:13]: typo: `PosePirorError` -> `PosePriorError`
We can also drop the error since it derives from cost and is implied. Eventually we will move this to embodied dir as well.

Comment (mhmukadam) [reply theseus/utils/examples/pose_graph/pose_prior_error.py:13]: Actually, is this needed? The functionality is similar to [variable_difference.py](https://github.com/facebookresearch/theseus/blob/main/theseus/embodied/misc/variable_difference.py).

Which maybe we should rename to Manifold difference if optim vars will be restricted to Manifold class later and add this to the related item in #100.

File: theseus/utils/examples/pose_graph/relative_pose_error.py [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+# LICENSE file in the root directory of this source tree.
+
+from typing import List, Optional, Tuple, Union
+
+import torch
+
+import theseus as th
+
+
+class RelativePoseError(th.CostFunction):
```
Comment (mhmukadam) [inline theseus/utils/examples/pose_graph/relative_pose_error.py:13]: [Between](https://github.com/facebookresearch/theseus/blob/main/theseus/embodied/measurements/between.py)?

**Review by mhmukadam — COMMENTED** [review]
Looks great. Left some comments.

File: theseus/utils/examples/pose_graph/relative_pose_error.py (hunk 3 above)
Comment (fantaosha) [reply theseus/utils/examples/pose_graph/relative_pose_error.py:13]: A little different. `Between` is `measurement^-1 * v0^-1 * v1` while `RelativePoseError` is `v1^-1*v0*measurement`. Also, it seems that the `jacobian` of `Between` is incorrect.

File: theseus/utils/examples/pose_graph/pose_prior_error.py (hunk 2 above)
Comment (fantaosha) [reply theseus/utils/examples/pose_graph/pose_prior_error.py]: It seems that the `jacobian` of `variable_difference` is incorrect. Maybe we need to fix this.

Comment (luisenp) [reply theseus/utils/examples/pose_graph/pose_prior_error.py]: Is it incorrect for some classes? What is the error?

File: theseus/utils/examples/pose_graph/relative_pose_error.py (hunk 3 above)
Comment (luisenp) [reply theseus/utils/examples/pose_graph/relative_pose_error.py:13]: Same question regarding the incorrect jacobian.

Comment (luisenp) [reply theseus/utils/examples/pose_graph/relative_pose_error.py:13]: Just saw the comments in the chat thread, we can continue the conversation there.

**Review by luisenp — COMMENTED** [review]
This looks good to me! Let's wait until we settle on a solution regarding the cost functions before merging. 

Also, we should add a version that includes an outer loss as well, although that can be done in a separate PR.

File: theseus/utils/examples/__init__.py [hunk 4]
```diff
@@ -40,3 +40,12 @@
         "Unable to import Tactile Pose Estimation utilities. "
         "Please make sure you have matplotlib and omegaconf installed."
     )
+
+
+try:
+    from .pose_graph import PosePriorError, RelativePoseError
+except ModuleNotFoundError:
+    warnings.warn(
+        "Unable to import Motion Planning utilities. "
```
[the remark this answers is not in the document]
Comment (fantaosha) [reply theseus/utils/examples/__init__.py:49]: I don't know. I definitely need to change the warning message.

[the remark this answers is not in the document]
Comment (luisenp) [reply theseus/utils/examples/__init__.py:49]: I think this should work without this try/except block, the other ones should take care of the previous issue.

File: theseus/utils/examples/pose_graph/dataset.py [hunk 5]
```diff
@@ -0,0 +1,151 @@
...
+        relative_pose: Union[th.SE2, th.SE3],
+        weight: Optional[th.DiagonalCostWeight] = None,
+    ):
+        self.i = i
+        self.j = j
+        self.relative_pose = relative_pose
+        self.weight = weight
+
+
+def read_3D_g2o_file(path: str) -> Tuple[int, List[th.SE3], List[PoseGraphEdge]]:
```
[the remark this answers is not in the document]
Comment (fantaosha) [reply theseus/utils/examples/pose_graph/dataset.py:30]: Added. See [here](https://github.com/facebookresearch/theseus/blob/2644d36f8dd71bbd6beaa0f094c37d6e95281e85/theseus/utils/examples/pose_graph/dataset.py#L28).

File: theseus/utils/examples/__init__.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (fantaosha) [reply theseus/utils/examples/__init__.py:49]: See #126

**Review by mhmukadam — APPROVED** [review]

**Comment by fantaosha** [comment]
> This looks good to me! Let's wait until we settle on a solution regarding the cost functions before merging.
> 
> Also, we should add a version that includes an outer loss as well, although that can be done in a separate PR.

Created an issue for this. See #127.
